contrib/src/compression.rs

```diff
@@ -0,0 +1,149 @@
+use brotli;
+#[cfg(feature = "brotli_compression")]
+use brotli::enc::backward_references::BrotliEncoderMode;
+
+#[cfg(feature = "gzip_compression")]
+use flate2;
+#[cfg(feature = "gzip_compression")]
+use flate2::write::GzEncoder;
+
+#[derive(Debug, Default)]
+pub struct Compression;
+
+impl Compression {
+    /// This function creates a Compression to be used in your Rocket
+    /// instance. Add ```.attach(rocket_contrib::Compression::fairing())```
+    /// to your Rocket instance to use this fairing.
+    ///
+    /// # Returns
+    ///
+    /// A Compression instance.
+    pub fn fairing() -> Compression {
+        Compression::default()
+    }
+
+    fn accepts_encoding(request: &Request, encodings: &[&str]) -> bool {
+        request
+            .headers()
+            .get("Accept-Encoding")
+            .flat_map(|e| e.split(","))
+            .any(|e| encodings.contains(&e.trim()))
+    }
+
+    fn set_body_and_header<'h, B>(response: &mut Response<'h>, mut body: B, header: &'h str)
+    where
+        B: Read + Seek + 'h,
+    {
+        if body.seek(SeekFrom::Start(0)).is_ok() {
+            response.remove_header("Content-Encoding");
+            response.adjoin_header(Header::new("Content-Encoding", header));
+            response.set_streamed_body(body);
+        }
+    }
+}
+
+impl Fairing for Compression {
+    fn info(&self) -> Info {
+        Info {
+            name: "Brotli and gzip compressors for responses",
+            kind: Kind::Response,
+        }
+    }
+
+    fn on_response(&self, request: &Request, response: &mut Response) {
+        let content_type = response.content_type();
+        // Images must not be compressed
+        if let Some(ref content_type) = content_type {
+            if content_type.top() == "image" {
+                return;
+            }
+        }
+
+        // The compression is done if the request supports brotli or gzip and
+        // the corresponding feature is enabled
+        if cfg!(feature = "brotli_compression")
+            && Compression::accepts_encoding(request, &["br", "brotli"])
+        {
+            let mut success = false;
+            let mut plain_buffer: Vec<u8> = Vec::new();
+            let mut compressed = Cursor::new(Vec::<u8>::new());
+            {
+                let plain = response.body();
+                if let Some(plain) = plain {
+                    let mut params = brotli::enc::BrotliEncoderInitParams();
+                    params.quality = 2;
+                    if let Some(ref content_type) = content_type {
+                        if content_type.top() == "text" {
+                            params.mode = BrotliEncoderMode::BROTLI_MODE_TEXT;
+                        } else if content_type.top() == "font" {
+                            params.mode = BrotliEncoderMode::BROTLI_MODE_FONT;
+                        }
+                    }
+                    if plain.into_inner().read_to_end(&mut plain_buffer).is_ok() {
```

Comment:
`read_to_end` is the problematic part of this: what if the response was streaming a 3GB file from disk? Without compression, it will be streamed to the client piece by piece keeping only a bit in memory at a time. With compression, it now has to read all 3 GB into memory, compress it (keeping both uncompressed and compressed versions in memory for a short time), then send the compressed version to the client.

A safe approach would involve something like this (modulo parameters and error handling):
```
let reader = response.take_body().into_inner();
let new_body = brotli::CompressorReader::new(reader, buffer_size, q, lgwin);
Compression::set_body_and_header(response, new_body, "br");
```

(The same reasoning applies to gzip below)

Reply:
I will try your suggestion when #646 is resolved and I can compile Rocket.

However, `flate2` API only supports to write `&[u8]`, so I don't find any manner of converting the `Read` into a `&[u8]` without `read_to_end` or similar (I guess that `take` has the same problem).

Thanks for your help.

Reply:
`flate2::read::GzEncoder` looks like it would work for this purpose  (instead of `flate2::write::GzEncoder`)

Reply:
I understand that then I could use `read` to write in a streamed manner the bytes compressed into a buffer of my desired size, no?

It seems weird to me that `read` writes compressed data xD, that is why I didn't found this API before.

Thanks again.